soroban-env-host/src/host.rs

```diff
@@ -1732,6 +1742,135 @@ impl VmCallerCheckedEnv for Host {
         self.create_contract_with_id_preimage(ScContractCode::Token, buf)
     }
 
+    //TODO: metering
+    fn create_token_wrapper(
+        &self,
+        vmcaller: &mut VmCaller<Host>,
+        asset: Object,
+    ) -> Result<Object, HostError> {
+        let a = self.visit_obj(asset, |hv: &Vec<u8>| {
+            //self.charge_budget(CostType::ValDeser, hv.len() as u64)?;
+            Asset::read_xdr(&mut hv.as_slice())
+                .map_err(|_| self.err_general("failed to de-serialize Asset"))
+        })?;
+
+        // TODO: Are these error msgs necessary? Should they be simplified?
+
+        // We want to convert xdr::Asset into an ScVal that matches the ClassicMetadata
+        // contracttype in the token contract
+        let arg =
+            match &a {
+                Asset::Native => ScVal::Object(Some(ScObject::Vec(ScVec(
+                    vec![ScVal::Symbol(
+                        "Native"
+                            .to_string()
+                            .try_into()
+                            .map_err(|_| self.err_general("invalid Native"))?,
+                    )]
+                    .try_into()
+                    .map_err(|_| self.err_general("invalid vec"))?,
+                )))),
+                Asset::CreditAlphanum4(a4) => {
+                    let PublicKey::PublicKeyTypeEd25519(issuer_key) = a4.issuer.0.clone();
+                    ScVal::Object(Some(ScObject::Vec(ScVec(
```

Comment:
This is the same block of code as below (modulo 4-12 diff), so could we refactor it into a helper function? It could either asset type for logging or just have generic error messages.